[Hub Generated] Review request for Microsoft.MachineLearningServices to update version preview/2020-09-01-preview

forteddyt opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/forteddyt/azure-rest-api-specs/thtodoro%2Fadd-mfe...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [x] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

harneetvirk committed:
Enabling java client (#1)

forteddyt committed:
Revert "Enabling java client (#1)" (#2)

This reverts commit 26873d33199c64732735a5c0f92fe90f4044453e.

forteddyt committed:
add mfe to 2020-09-01-preview

forteddyt committed:
add to custom-words

forteddyt committed:
add 200 to async delete response

forteddyt committed:
change default tag to package-2020-09-01-preview

forteddyt committed:
[TEST] remove jobType discriminator/requirement from ComputeJobBase

forteddyt committed:
[TEST] modify desc

forteddyt committed:
Revert "[TEST] modify desc"

This reverts commit 22fcc0bc05f731da606afa4f14621508b4d66fd9.

forteddyt committed:
Revert "[TEST] remove jobType discriminator/requirement from ComputeJobBase"

This reverts commit c09fdec1a895ee0c83c6cbd935978cf74f2d631e.

forteddyt committed:
[TEST] try 2020-12 spec

forteddyt committed:
Revert "[TEST] try 2020-12 spec"

This reverts commit 4682d398d4cf0a91d1dd1310be74112f94f08e3b.

forteddyt committed:
move java portion to own readme file

forteddyt committed:
updated sdk-for-net to -track2

forteddyt committed:
to trigger gates

harneetvirk committed:
Enabling Java client (#6)

harneetvirk committed:
Adding 2020-09-01-preview for Go (#7)

* Enabling Java client

* Enabling Go client for 2020-09-01-preview

* Removing track2 for .net

forteddyt committed:
Merge remote-tracking branch 'upstream/master'

allenjzhang commented:
/azp run SDK azure-sdk-for-net

allenjzhang commented:
/azp run

harneetvirk commented:
/azp run

forteddyt committed:
add top-level consumes/produces props to spec

forteddyt committed:
add missing security top-level prop

forteddyt committed:
Merge branch 'master' into thtodoro/add-mfe

forteddyt committed:
update MFE to latest PuP spec

forteddyt committed:
update examples + prettier

forteddyt committed:
re-add LabelingJob, remove CommandJob/poll, prettier

forteddyt committed:
remove SweepJob poll

forteddyt committed:
remove internal x-ms-discriminator-value prop

forteddyt committed:
Merge remote-tracking branch 'upstream/master' into thtodoro/add-mfe

forteddyt committed:
remove jobs.json, since mfe already has their definitions

forteddyt committed:
add L+ingJobs ExportSummary api + ui; +missing discriminator definitions; +update to latest mfe

forteddyt committed:
update examples

forteddyt committed:
reference mfe.json instead of jobs.json in azureresourcesschema.md

harneetvirk commented:
/azp run

forteddyt committed:
test sweepjob Body and ParameterSamplingConfiguration fix

harneetvirk commented:
/azp run

forteddyt committed:
request bodies should be $ref

forteddyt committed:
re-encompass ParameterSamplingConfiguration in allOf

forteddyt committed:
flatten only azure resource's property bag

forteddyt committed:
flatten JobBase

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CodeContainer"
+            }
+          ],
```
Anonymous `allOf` is not recommended
```suggestion
              "$ref": "#/definitions/CodeContainer",
```

forteddyt replied:
OpenApi 2.0 spec specifies that sibling properties will be ignored if `$ref` is present - https://github.com/OAI/OpenAPI-Specification/issues/1514. We need `"x-ms-client-flatten": true` and (to a lesser extent) the description properties, so we encompass all $references with AllOf

ArcturusZhang replied:
Despite the OpenApi has this restriction, but autorest treats this case differently. Autorest could properly identify the description here. @PhoenixHe-msft to confirm

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+          "type": "string"
+        },
+        "innerError": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/InnerErrorResponse"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/InnerErrorResponse"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+          }
+        },
+        "innerError": {
+          "allOf": [
+            {
+              "$ref": "#/definitions/InnerErrorResponse"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/InnerErrorResponse"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "assetPath": {
+          "description": "Details of asset path in the datastore.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AssetPath"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/AssetPath"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CodeVersion"
+            }
+          ],
```
```suggestion
              "$ref": "#/definitions/CodeVersion",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DataContainer"
+            }
+          ],
```
```suggestion
              "$ref": "#/definitions/DataContainer",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "accountKey": {
+          "description": "Storage account key authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AccountKeySection"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/AccountKeySection"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "certificate": {
+          "description": "Service principal certificate authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/CertificateSection"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/CertificateSection"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "sas": {
+          "description": "Storage container SAS token authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/SasSection"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/SasSection"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "servicePrincipal": {
+          "description": "Service principal password authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ServicePrincipalSection"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/ServicePrincipalSection"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "sqlAdmin": {
+          "description": "SQL user/password authentication.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/SqlAdminSection"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/SqlAdminSection"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "properties": {
+        "credentials": {
+          "description": "Azure Data Lake credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "properties": {
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "properties": {
+        "credentials": {
+          "description": "Azure SQL database credentials.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/DatastoreCredentials"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/DatastoreCredentials"
```

ArcturusZhang reviewed (changes requested):
We have a lot of anonymous allOf in this swagger, which is highly not recommended.
This is the first round of review... to be continued...

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "additionalProperties": false
+    },
+    "CocoExportSummary": {
+      "allOf": [
+        {
+          "$ref": "#/definitions/ExportSummary"
+        },
+        {
+          "type": "object",
+          "properties": {
+            "snapshotPath": {
+              "description": "The output path where the labels will be exported.",
+              "type": "string",
+              "readOnly": true
+            },
+            "containerName": {
+              "description": "The container name to which the labels will be exported.",
+              "type": "string",
+              "readOnly": true
+            }
+          }
+        }
+      ],
```
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "snapshotPath": {
              "description": "The output path where the labels will be exported.",
              "type": "string",
              "readOnly": true
            },
            "containerName": {
              "description": "The container name to which the labels will be exported.",
              "type": "string",
              "readOnly": true
            }
          },
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "additionalProperties": false
+    },
+    "CsvExportSummary": {
+      "allOf": [
+        {
+          "$ref": "#/definitions/ExportSummary"
+        },
+        {
+          "type": "object",
+          "properties": {
+            "snapshotPath": {
+              "description": "The output path where the labels will be exported.",
+              "type": "string",
+              "readOnly": true
+            },
+            "containerName": {
+              "description": "The container name to which the labels will be exported.",
+              "type": "string",
+              "readOnly": true
+            }
+          }
+        }
+      ],
```
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "snapshotPath": {
              "description": "The output path where the labels will be exported.",
              "type": "string",
              "readOnly": true
            },
            "containerName": {
              "description": "The container name to which the labels will be exported.",
              "type": "string",
              "readOnly": true
            }
          },
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "additionalProperties": false
+    },
+    "DatasetExportSummary": {
+      "allOf": [
+        {
+          "$ref": "#/definitions/ExportSummary"
+        },
+        {
+          "type": "object",
+          "properties": {
+            "labeledAssetName": {
+              "description": "The unique name of the labeled data asset.",
+              "type": "string",
+              "readOnly": true
+            }
+          }
+        }
+      ],
```
```suggestion
      "allOf": [
        {
          "$ref": "#/definitions/ExportSummary"
        }
        ],
          "type": "object",
          "properties": {
            "labeledAssetName": {
              "description": "The unique name of the labeled data asset.",
              "type": "string",
              "readOnly": true
            }
          },
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ModelContainer"
+            }
+          ],
```
```suggestion
              "$ref": "#/definitions/ModelContainer",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "assetPath": {
+          "description": "Details of asset path in the datastore.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/AssetPath"
+            }
+          ]
```
```suggestion
              "$ref": "#/definitions/AssetPath"
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "properties": {
+          "description": "Additional attributes of the entity.",
+          "allOf": [
+            {
+              "$ref": "#/definitions/ModelVersion"
+            }
+          ],
```
```suggestion
              "$ref": "#/definitions/ModelVersion",
```

ArcturusZhang commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+        },
+        "systemData": {
+          "description": "System data associated with resource provider",
+          "allOf": [
+            {
+              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData"
+            }
+          ],
```
```suggestion
              "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
```

ArcturusZhang reviewed (changes requested):
Second part of anonymous `allOf` issues...

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into thtodoro/add-mfe

Commit:
remove allOf wrap on AnonymousAllOfs

Commit:
remove Labeling discriminator from inner property

forteddyt commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+      "required": [
+        "dockerImageUri"
+      ],
+      "allOf": [
+        {
+          "$ref": "#/definitions/DockerSpecification"
+        },
+        {
+          "type": "object",
+          "properties": {
+            "dockerImageUri": {
+              "description": "Image name of a custom base image.\r\n<seealso href=\"https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-custom-docker-image#use-a-custom-base-image\" />",
+              "type": "string"
+            }
+          }
+        }
+      ],
```
Can I get an explanation on why this is preferred? AllOf is supposed to take in an [array of object definitions](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#composition-and-inheritance-polymorphism) that compose a single object. This is the intention of what we are doing with our AllOfs

forteddyt replied:
@ArcturusZhang

Commit:
add OnlineEndpoint and OnlineDeployment to spec

forteddyt commented:
Ready for re-review, @ArcturusZhang. Also could you please address my questions?

Commit:
remove required prop from PATCH

Commit:
remove Error from resource responses

Commit:
AuthKeys -> EndpointAuthKeys; AuthToken -> EndpointAuthToken

Commit:
OSType -> OSTypes

Commit:
remove x-ms-secret prop from EndpointAuthKeys props

Commit:
only references should be wrapped in AllOf in definitions

forteddyt commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -0,0 +1,6276 @@
+            }
+          }
+        }
+      ],
```
you're right, but I will ignore missing descriptions for now

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into thtodoro/add-mfe

Commit:
reference common type api version, sub id, and rg name params

Commit:
Revert "reference common type api version, sub id, and rg name params"

This reverts commit affb78d06e9ed81ff99579836e32170aebd93f11.

forteddyt commented:
All future new specs will reference common types. Reverting references on this 2020-09-01-preview

forteddyt committed:
update spec to latest MFE

forteddyt committed:
update examples

forteddyt committed:
remove 'Properties' suffix from OE/OD

forteddyt committed:
update examples

forteddyt committed:
remove ReadOnly prop from TrackedResource's Location

forteddyt committed:
update to latest mfe + add PATCH to OD

forteddyt committed:
typo fix

forteddyt committed:
prettier

forteddyt committed:
ProvisioningState should be ReadOnly

ArcturusZhang commented:
Hey, do we have any update on this PR?

forteddyt committed:
remove misleading default for async headers

forteddyt committed:
update to latest MFE

forteddyt committed:
example update

forteddyt committed:
add AutoMlJob and PipelineJob

forteddyt committed:
add ComponentContainer + ComponentVersion

forteddyt committed:
add ComponentContainer + ComponentVersion examples

forteddyt commented:
hey, sorry for the delay. Yes, in addition to updating the spec with the latest (minor) changes, we added AutoMlJob and PipelineJob to the existing Job API. We also added a new Component API, that has ComponentContainer and ComponentVersion.

Could these new additions be reviewed?

forteddyt committed:
update custom-words

forteddyt commented:
I'm (expectedly) getting the same LintDiff issues in this PR as in my [2021-03-01-preview PR](https://github.com/Azure/azure-rest-api-specs/pull/12554), so I shall voice my questions/concerns here:

1. Regarding [R2008](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R2008), I was talking with another ARM reviewer last week and they were saying the issue was that the `$ref` definition itself wasn't `readOnly: true`, but the definition that used the reference was `readOnly: true`. To that, I'm wondering: why is that an issue? What if a `$ref` definition is used as `readOnly: true` in one definition, but is mutable in another? for example: 
```json
"output": {
   "ref" :"#/definitions/Job",
   "readOnly": true,
   "x-ms-mutability": ["read"]
}
"input": {
   "ref": "#/definitions/Job",
   "readOnly": false,
   "x-ms-mutability": ["read", "write"]
}
```
2. Regarding [R2026](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R2026), in our codebase this is represented as a `Dictionary<string, Dictionary<string, string>>`. If the current definition is truly incorrect, I'm having trouble understanding how to properly represent this in the spec... Could I get some guidance on this?
3. Regarding [R3015](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R3015), I was told by a previous ARM reviewer to ignore this. "enum" is the name of the property, not an Enum-type property

ArcturusZhang commented:
CC @jianyexi for the Linter questions

forteddyt committed:
add Schedule API

forteddyt committed:
add Schedule examples

forteddyt commented:
Additionally, we added the Schedule API

forteddyt commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:
```diff
@@ -9788,6 +10010,301 @@
+            "modelAsString": true
+          }
+        },
+        "provisioningStatus": {
```
This is **not** meant to be the `provisioningState` property used in asynchronous operations, as defined [here](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#provisioningstate-property). Will the name similarity be an issue?

forteddyt committed:
reference enums by definition

forteddyt committed:
reference MLS' ComputeType definition

forteddyt committed:
Revert "reference MLS' ComputeType definition"

This reverts commit f672037b7cf7e96f27f302dab21deae519fb475b.

forteddyt committed:
Schedule now uses ScheduleComputeType

pilor commented:
>         "createdTimeUtc": {

This should be in systemData.  You shouldn't duplicate it inside the resource properties

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8568 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

pilor commented:
>     "ResourceIdentityInArm": {

please mark all managed identity properties readonly that are readonly

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

pilor commented:
>         "clientSecretUrl": {

this is not something you expose to the user.  Please read through the managed identity implementation/contract docs one more time and ensure your contract is correct

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9085 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

pilor commented:
>         "type": {

Don't use "type" as the discriminator on all these definitions.  Change it to something more specific like "scheduleType" for schedules, etc...  You got exceptions to use "tags", etc... within the property bag but there is no reason to duplicate all top level property names inside your property bag when you can easily use something else

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:10184 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

forteddyt committed:
remove Schedule API

forteddyt committed:
remove Schedule API examples

forteddyt committed:
remove LabelingJob's createdTimeUtc

forteddyt committed:
rename 'type' properties

forteddyt committed:
remove mfeModel.out

forteddyt committed:
remove clientSecretUrl

forteddyt commented:
> > ```
> > "ResourceIdentityInArm": {
> > ```
> 
> please mark all managed identity properties readonly that are readonly
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in [9f78980](https://github.com/Azure/azure-rest-api-specs/commit/9f789805b324a5ab3b788475f6ae72ed0f55ce8e). [](commit_id = [9f78980](https://github.com/Azure/azure-rest-api-specs/commit/9f789805b324a5ab3b788475f6ae72ed0f55ce8e), deletion_comment = False)

From my understanding, ManagedIdentities are updatable. Our service supports updating them, as well

forteddyt committed:
renamed IdAssetReference's id -> assetId

forteddyt committed:
rename ResourceIdentityInArm's type to identityType

pilor commented:
>         "identityType": {

"type" was the correct property name

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8857 in 3c977ff. [](commit_id = 3c977ff0379a9c1251401f977b78fd2f715eb728, deletion_comment = False)

pilor commented:
>     "ResourceIdentityInArm": {

"principalId" and "tenantId" still need to be marked readonly

---
In reply to: [808519549](https://github.com/Azure/azure-rest-api-specs/pull/12508#issuecomment-808519549) [](ancestors = 808519549)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:9097 in 9f78980. [](commit_id = 9f789805b324a5ab3b788475f6ae72ed0f55ce8e, deletion_comment = False)

pilor commented:
>           "type": "string"

"resourceId" and "tenantId" are not part of the contract.  See [https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response](https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8849 in 3c977ff. [](commit_id = 3c977ff0379a9c1251401f977b78fd2f715eb728, deletion_comment = False)

forteddyt committed:
Merge branch 'master' of github.com:Azure/azure-rest-api-specs

forteddyt committed:
Merge branch 'master' into thtodoro/add-mfe

forteddyt committed:
ResourceIdentityInArm -> ResourceIdentity

forteddyt commented:
> > ```
> >       "type": "string"
> > ```
> 
> "resourceId" and "tenantId" are not part of the contract. See https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingUserAssigned.html#resource-definition-response
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/mfe.json:8849 in [3c977ff](https://github.com/Azure/azure-rest-api-specs/commit/3c977ff0379a9c1251401f977b78fd2f715eb728). [](commit_id = [3c977ff](https://github.com/Azure/azure-rest-api-specs/commit/3c977ff0379a9c1251401f977b78fd2f715eb728), deletion_comment = False)

I've updated ResourceIdentityInArm's definition (alongside renaming it to ResourceIdentity) in c5e2292. Could you take a look and tell me if it's properly defined? I'll update the example files afterwards

forteddyt committed:
add missing comma to SystemAssignedUserAssigned

forteddyt committed:
ResourceIdentityType -> ResourceIdentityAssignment to avoid conflict

forteddyt committed:
Merge remote-tracking branch 'upstream/master'

forteddyt committed:
Merge remote-tracking branch 'upstream/master'

forteddyt committed:
Merge branch 'master' into thtodoro/add-mfe

ArcturusZhang commented on specification/machinelearningservices/resource-manager/readme.md:
````diff
@@ -25,6 +25,10 @@ To see additional help and options, run:
 These are the global settings for the Machine Learning Services API.
 
 ``` yaml
+<<<<<<< HEAD
+openapi-type: arm 
+tag: package-2020-09-01-preview
+=======
````
Hey, looks like that we have some conflicts that are not correctly resolved here
